[Stdlib] Improves Collection.sort  to accept throwing closure

codestergit opened the pull request:
<!-- Please complete this template before creating pull request. -->
#### What's in this pull request?

This pull request improves `MutableCollectionType.sort` ,`MutableCollectionType.sortInPlace` and other sort related functions to take throwing clousre.
#### Resolved bug number: ([SR-715](https://bugs.swift.org/browse/SR-715))

<!-- If this pull request resolves any bugs from Swift bug tracker -->

---

<!-- This selection should only be completed by Swift admin -->

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>



The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

<!-- Thank you for your contribution to Swift! -->

This commit resolves https://bugs.swift.org/browse/SR-715

codestergit commented:
There are two tests which are failing.
1) ` test/DebugInfo/closure-multivalue.swift`
I have tried to genereate `IR` and with my changes it is showing `linkageName: _TPA__TTRXFo_oSSoSS_dSb_XFo_iSSiSS_dSb_` previously it was ``linkageName: _TTRXFo_oSSoSS_dSb_XFo_iSSiSS_dSb_`. So I have changed but it did not pass the testCase.
Here:
https://github.com/apple/swift/pull/1527/files#diff-fe533918e3617168eef57e89f39cfe2fL37
Please let me know how to resolve this and what I am missing.
2) `test/SourceKit/SourceDocInfo/cursor_stdlib.swift`
I have tried to [fix](https://github.com/apple/swift/pull/1527/files#diff-a04c3ccbb871153dc04738fdd1b3730cL36) this but it is failing.
I need help to fix these test cases.
Thanks

gribozavr commented:
@codestergit Thank you!  Two requests from me.

1.  Please add documentation about the behavior of this method when an error is thrown.  (It actually has mutated the collection, but only reordered the elements.  No elements are lost, no new elements are introduced.)

2.  The tests you are added are a good start, but we can improve.  Please take a look at existing tests here `stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift.gyb` and try to come up with a few cases that sort a collection and the comparator throws an error in the middle of the operation.  Then check that the sorting kept the same set of elements in the collection.  Please add new tests to that common file, so that the tests get run for every collection implementation, increasing our coverage.

videmort commented:
Please Fix conflicts

codestergit commented:
There are lot of conflicts because of swift 3 changes. I am redoing my changes.I will merge the changes soon by addressing comments :)

videmort commented:
I'm Happy to hear that!!

2016-03-22 11:27 GMT+01:00 codester <notifications@github.com>:

> There are lot of conflicts because of swift 3 changes. I am redoing my
> changes.I will merge the changes soon by addressing comments :)
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/apple/swift/pull/1527#issuecomment-199741215>
>

codestergit commented:
First of all I am really sorry for the delay. I got some urgent work and need to go out of station so this got delay.
I have updated the pull request and test are passing but I got one problem. `sorted` is working as expected. When predicate `isOrderedBefore` throws in between sorting it will return nothing and source array remains unchanged.
But `sort` is mutating the collection and if  `isOrderedBefore` throws in between sorting there may be case original source elements may be lost as `insertionSort` algo not able to complete it's loop and swap so elements may get lost or duplicate. I have commented the same test case in (ErrorHandling.swift test)[https://github.com/apple/swift/pull/1527/files#diff-474e7be2f2a88d7103e1fef56d517f46R343] .
Here I have given [3, 2, 1] and  if  `isOrderedBefore` throws in between the `source` array gets mutated and changed to [2, 3, 3].
We can not catch the error and throw again as it have to `rethrows` and I do not know it is best idea to save the collection copy.
@gribozavr  @dabrahams any suggestions how to handle this or is this required behaviour??
Thanks for your help as always :)

codestergit commented on test/1_stdlib/ErrorHandling.swift:
```diff
@@ -340,4 +340,42 @@ ErrorHandlingTests.test("ErrorHandling/Collection map") {
+      print("collection before error \(collection)")
+      print("a \(a) \(b)")
+
+      if b == 2 {
```
With this case `sort` mutates the collection to [2, 3, 3]. `insertionSort` not able to complete the loop and swap elements so we are loosing the elements from original collection.

dabrahams replied:
This behavior is OK.  See https://en.wikipedia.org/wiki/Exception_safety

gribozavr replied:
I find losing elements surprising though.  Wouldn't we want to ensure that we don't lose elements as a matter of QoI for this API specifically?

codestergit replied:
@gribozavr It throws [here](https://github.com/codestergit/swift/blob/master/stdlib/public/core/Sort.swift.gyb#L75) and swap operation not able to complete. We also do partition of elements if elements are above 20. It may duplicate elements and loose.

gribozavr replied:
I see.  Can we catch the error, store the elements back in arbitrary order, and re-throw?

codestergit replied:
But the function has to be `rethrows`. If we throw error again then we need to define one another `sorted throw` function which throw the error again. One another solution might be we can make copy of original collection and mutate collection if sorting is successful else discard the collection. So original collection will not be changed. But it will take extra space.

gribozavr replied:
> If we throw error again then we need to define one another sorted throw function which throw the error again.

This works:

```func foo(e: () throws -> ()) rethrows {
  do {
    try e()
  } catch {
    throw error
  }
}
```

> One another solution might be we can make copy of original collection

I don't think that kind of performance kit would be acceptable.

codestergit replied:
It will give error. I have tried that. Error: "A function declared rethrows may only throw if  parameter does". 
If I make some other function say `_insertionSort` to `throws` instead of `rethrows` than it will give error while calling `try _insertionSort(&elements, range , isOrderedBefore)` // Error: "Call can throw but error is nit handled; a function decalred rethrows may only throw if parameter does". It might be case we are calling `sorted` without throwing clousre.

gribozavr replied:
> Error: "A function declared rethrows may only throw if parameter does". 

Could it be because we are generating two overloads of _insertionSort, one that calls a throwing predicate, and another one that calls `<` that can't throw?

codestergit replied:
I think because previously defined `foo` can be called as `foo() {}` or `foo(e: {
throw error
}`
```swift
func foo(e: () throws -> ()) rethrows {
  do {
    try e()
  } catch {
    throw error
  }
}
```
above function will not compile. Am I missing something ??

gribozavr replied:
Which compiler are you using?  It works for me in a REPL:

```
$ ./bin/swift
***  You are running Swift's integrated REPL,  ***
***  intended for testing purposes only.       ***
***  The full REPL is built as part of LLDB.   ***
***  Type ':help' for assistance.              ***
(swift) func foo(e: () throws -> ()) rethrows {
            do {
              try e()
          } catch {
              throw error //Giving error: "A function declared rethrows may only throw if parameter does".
          }
        }
(swift) foo { print("hello") }
hello
```

rjmccall replied:
That was a relatively recent improvement, but I assume that anybody hacking on trunk has it.

codestergit replied:
I was trying it on playground `Swift 2.2`. Thanks I will try it on latest master.

dabrahams replied:
on Thu Apr 21 2016, Dmitri Gribenko <notifications-AT-github.com> wrote:

>
> I find losing elements surprising though. Wouldn't we want to ensure that we
> don't lose elements as a matter of QoI for this API specifically?

No, trying to do that is a losing battle and there's nothing useful the
user can do with the result anyway.

-- 
Dave

codestergit replied:
Ok than I will update the pull request with test cases for `sorted`. Thanks @dabrahams  and @gribozavr

gribozavr replied:
We discussed this with @dabrahams and we agreed that it is a good idea to do this as QoI, but not document the guarantee.

jrose-apple replied:
I know I'm arriving late, but there is *plenty* they can do with the result. In particular, they might re-sort using a non-throwing sorting algorithm (perhaps a more expensive one).

dabrahams replied:
Please describe the details of a realistic example where it makes sense to code these kinds of fallbacks into your app.

jrose-apple replied:
If your comparator can throw, you're forced to handle that *somehow.* I can't come up with a realistic example of a comparator that can throw (i.e. I'm not sure why we're doing this at all), so I can't come up with a realistic example of a fallback. But as soon as I write that `try` I have to decide what to do about it.

dabrahams replied:
Yes, and in 99% of cases a partly sorted collection is useless to you and you have no alternative way to do the same sort, so in 99% of cases you’ll be sorting something you’re willing to lose, like a copy of the original collection.  

By the way, if you *did* have a fallback for a failing comparison, you’d just make your comparison fall back to it rather than aborting the sort and starting over again.

jrose-apple replied:
That's what I was thinking too, and so I couldn't think of why you'd want to have a throwing comparator at all. But I can imagine wanting to have the fallback be consistent: either everything sorts using the "good" comparison, or everything sorts using the fallback.

dabrahams replied:
Yes, this is all theoretically possible but highly unlikely and not worth designing for.  It’s reasonable to do something as QOI but the precedent set by documenting the behavior would significantly complicate what people need to do to reason about (and document their own) error handling.

jrose-apple replied:
If it's not worth designing for, I'm not sure why it's worth implementing. If it's there, someone will start depending on it.

dabrahams replied:
on Fri Apr 22 2016, Jordan Rose <notifications-AT-github.com> wrote:

> In test/1_stdlib/ErrorHandling.swift:
>
>> @@ -340,4 +340,42 @@ ErrorHandlingTests.test("ErrorHandling/Collection map") {
>>    }
>>  }
>>  
>> +/* when b == 2 we are losing elements. insertion sort throws error
>> +in between while sorting so loop is not able to complete and swap the elements and array mutates
>> +to [2,3,3]
>> +ErrorHandlingTests.test("ErrorHandling/sort") {
>> +  var collection = [3, 2, 1]
>> +  do {
>> +    try collection.sort { (a, b) throws -> Bool in
>> +      print("collection before error \(collection)")
>> +      print("a \(a) \(b)")
>> +
>> +      if b == 2 {
>
> If it's not worth designing for, I'm not sure why it's worth implementing. If
> it's there, someone will start depending on it.

It will prevent one or two naive users from having a subtle bug, it will
prevent us from fielding/rejecting a long stream of bug reports from
people who think they should be able to rely on it, and I don't mind
people relying on it.  What I don't want, at least not without
significantly more consideration, is to formalize this as part of how we
describe error handling behavior.

-- 
Dave

dabrahams commented:
on Thu Apr 21 2016, codester <notifications-AT-github.com> wrote:

> First of all I am really sorry for the delay. I got some urgent work and need to
> go out of station so this got delay.
> I have updated the pull request and test are passing but I got one problem.
> sorted is working as expected. When predicate isOrderedBefore throws in between
> sorting it will return nothing and source array remains unchanged.
> But sort is mutating the collection and if isOrderedBefore throws in between
> sorting there may be case original source elements may be lost as insertionSort
> algo not able to complete it's loop and swap so elements may get lost or
> duplicate. 

There's nothing wrong with that.  The only requirement of an operation
that throws, unless we explicitly add stronger requirements, is that it
doesn't break any locally-known invariants.  This is called the “basic
guarantee.” https://en.wikipedia.org/wiki/Exception_safety


-- 
Dave

codestergit commented:
@dabrahams  @gribozavr   Updated the pull request. Please review it.

I have caught the error [here](https://github.com/apple/swift/pull/1527/commits/a698c4067ffd17e1891f56c3e0fe5fcd474c032b#diff-68c134d8478193cbe9299610501ff45bR79) and thrown it again. I have tried to test this thoroughly and it passed the tests. [We do partition and sort with greater than 20 elements.](https://github.com/apple/swift/pull/1527/commits/a698c4067ffd17e1891f56c3e0fe5fcd474c032b#diff-68c134d8478193cbe9299610501ff45bL180). So checking all permutations with more than 20 elements was not feasible so I changed the algorithm locally and tried to do partition and sort with all permutations of 5 element sequence and it passed all the tests.
Please let me know if any changes required.
Thanks!

CodaFi commented:
@codestergit I'm sorry that this fell by the wayside.  If you can rebase and fix the merge conflicts could you please do so.  Else I can take over this PR.

codestergit commented:
@CodaFi  Thanks, I will rebase and update the pull request.

codestergit commented:
@CodaFi  @gribozavr 
Updated the pull request. Please review the pull request.
Please let me know if any changes required.
Thanks

dabrahams commented on stdlib/public/core/CollectionAlgorithms.swift.gyb:
```diff
@@ -269,6 +269,9 @@ extension ${Self} {
   /// predicate as the comparison between elements.
   ///
+  /// This method can take throwing closure. If closure throws error while 
+  /// sorting, it will return empty collection.
```
This seems like the wrong behavior.  Why isn't the error simply propagated outward (in which case nothing is actually returned)?

codestergit replied:
Sorry I made mistake here. It will throw error. I will correct it. 
Thanks

codestergit replied:
@dabrahams Removed comment. Thanks.
Please let me know if any other changes required.

dabrahams commented on stdlib/public/core/Sort.swift.gyb:
```diff
@@ -53,10 +62,23 @@ func _insertionSort<C>(
+            break
+          }
+        } catch {
+          elements[i] = x
```
Why bother doing the above?  Isn't `elements` about to be discarded?

codestergit replied:
No, If we are mutating a collection with `sort(by:)` than collection has been changed and at this line we are trying to swap the elements.
For eg: if I have array
 var a = [ 5, 4, 3 ,2, 1]
if we throw at  b == 4 than sort(by:) mutate the collection to
[4, 5, 5, 2, 1]
as the [last step](https://github.com/apple/swift/pull/1527/files/e86fb6e14bbf463f6b591042e52e6c50e6130392#diff-68c134d8478193cbe9299610501ff45bR88)
not able to swap the element and we thrown in between. So that is why we are catching the error and throwing it again. 
It just take care of no elements will lost and add.

dabrahams commented:
on Sun Jan 29 2017, codester <notifications-AT-github.com> wrote:

> So that is why we are catching the error and throwing it again.  It
> just take care of no elements will lost and add.

OK, that is a nice QOI feature for algorithms that only permute the
elements (provided it doesn't hurt performance), but keep in mind that
in general it's not something we can guarantee for mutating algorithms
that rethrow.

codestergit commented:
>  but keep in mind that
> in general it's not something we can guarantee for mutating algorithms
> that rethrow.

@dabrahams Yes, I agree with you. Previously @gribozavr suggested that we should try to handle this. If you would like I can remove this [documentation].

>  No elements will be lost.

(https://github.com/apple/swift/pull/1527/files/e86fb6e14bbf463f6b591042e52e6c50e6130392#diff-f87017c0446dd290ac092ef4901bd463R399)

airspeedswift commented:
This PR conflicts with @natecook1000’s one (#6638) to de-gyb sorts. Does it make sense to combine them or should one go before the other?

airspeedswift commented:
@swift-ci Please test

codestergit commented:
@airspeedswift Why 

> Wrong sha e86fb6e14bbf463f6b591042e52e6c50e6130392

Error by build bot.

dabrahams commented:
@airspeedswift I agree with @natecook1000

dabrahams commented:
on Sun Jan 29 2017, codester <notifications-AT-github.com> wrote:

> @dabrahams Yes, I agree with you. Previously @gribozavr suggested that
> we should try to handle this. If you would like I can remove this
> [documentation].

No, the docs are good.  I remember my discussion with @gribozavr and it
matches what you're telling me.  Thanks for putting up with my
forgetfulness!

-- 
-Dave

airspeedswift commented:
@dabrahams you mean you agree with @codestergit ? Or is there a reply from @natecook1000 I missed?

airspeedswift commented:
@codestergit re the tests: I think those failures are bogus reports from a previous failure that reported once I triggered a fresh test. Looks like the OS X build succeeded. There's an unrelated problem with Linux PR testing, once that's fixed we should be good to retest and then merge.

codestergit commented:
Thanks!! @airspeedswift  and @dabrahams

natecook1000 commented:
Happy to see this go in first—the degybification may have a performance regression and is NFC, so if this ready to go we should merge it.

Speaking of which…

natecook1000 commented:
@swift-ci Please benchmark

airspeedswift commented:
So.... some of those benchmarks are just noise (like MapReduceAnyCollection and the Objc bridging stuff), but some of them look like definite regressions resulted. For example RGBHistogram does a lot of sorting and has been very hard hit.

codestergit commented:
@airspeedswift @dabrahams @natecook1000 
If I am not passing throwing closure it should affect that much  😢.  

Do you think this [do/catch](https://github.com/apple/swift/pull/1527/files#diff-68c134d8478193cbe9299610501ff45bR74)  might affect the performance.

As previously @dabrahams has said that it might affect performance. Should I try to remove this and we can benchmark again? (It might fail validation tests I have written)

codestergit commented:
I checked the test cases of [RGBHistogram](https://github.com/apple/swift/blob/adc54c8a4d13fbebfeb68244bac401ef2528d6d0/benchmark/single-source/RGBHistogram.swift) at all places we are using `sorted()` without clousre. So It will call `_insertionSort` without `areInIncreasingOrder` clousre. 
Which is exactly same flow was calling (before this pull request). It is very strange their time got exceeded. Am I missing something here?
🤔

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
on Mon Jan 30 2017, Ben Cohen <notifications-AT-github.com> wrote:

> @dabrahams you mean you agree with @codestergit ? 

Yes, that's what I mean; the original was fine and I'm not requesting
any change.

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
on Tue Jan 31 2017, codester <notifications-AT-github.com> wrote:

> I checked the test cases of
> [RGBHistogram](https://github.com/apple/swift/blob/adc54c8a4d13fbebfeb68244bac401ef2528d6d0/benchmark/single-source/RGBHistogram.swift)
> at all places we are using `sorted()` without clousre. So It will call
> `_insertionSort` without `areInIncreasingOrder` clousre.
> Which is exactly same flow was calling (before this pull request). It is very strange their time got
> exceeded. Am I missing something here?
> 🤔

You might take a look at the generated code.  If it's still mysterious
after that we can ask some optimization experts to take a look at it.

-- 
-Dave

codestergit commented:
@dabrahams I am really sorry. I overlooked the code of  [RGBHistogram](https://github.com/apple/swift/blob/adc54c8a4d13fbebfeb68244bac401ef2528d6d0/benchmark/single-source/RGBHistogram.swift)

Noticed that it was [clousre](https://github.com/apple/swift/blob/adc54c8a4d13fbebfeb68244bac401ef2528d6d0/benchmark/single-source/RGBHistogram.swift#L151) and use sorted(by:) function.

airspeedswift commented:
If we do end up getting to the point where we need to debug generated code for performance, this makes it more worthwhile to combine this + the de-gybbed PR, since it also had unexpected performance implications.

airspeedswift commented:
Hi @codestergit – the hope is that PR #7672 should resolve some of these performance issues. Do you think you could rebase to pull it in, then we can rebenchmark?

codestergit commented:
@airspeedswift  Sure I will do it by tomorrow. 
Thanks

codestergit commented:
@airspeedswift @dabrahams  Rebased the pull request. Please run the benchmark. 

I have commented the [do/catch block](https://github.com/apple/swift/pull/1527/commits/03b28f87bf3c8e0c99347fa5c0f1518dc133c9e6) . I will uncomment it after benchmarks to see if it is impacting performance. Some of the validation test case will fail because of this.
Thanks

airspeedswift commented:
@swift-ci Please benchmark

airspeedswift commented:
Much better! @codestergit want to try putting that do/catch back in?

codestergit commented:
@airspeedswift @dabrahams  Great!!
Updated the pull request. Please run the benchmark again.

rintaro commented:
@swift-ci Please benchmark

codestergit committed:
[Stdlib] Improves sort and sorted to accept throwing clousre

This commit resolves https://bugs.swift.org/browse/SR-715

rintaro commented:
@swift-ci Please benchmark

codestergit commented:
@airspeedswift  @dabrahams 😁
Benchmark results looks good.
I think https://github.com/apple/swift/pull/7672 resolved the performance issue. It seems to just noise. 

Please let me know if you need any other changes. 
Thanks

rintaro commented:
Let's run full test.
@swift-ci Please test

rintaro commented:
Linux test failed with unrelated test case (Syntax)
@swift-ci Please smoke test Linux platform

codestergit commented:
@airspeedswift 
Why linux test is failing. Is it because of my changes?

codestergit committed:
Fix the testcase

codestergit commented:
@dabrahams Updated the pull request to resolve conflicts. 
The benchmark looks good. Can you please review and merge it.
Thanks

natecook1000 commented:
@swift-ci Please smoke test

dabrahams commented:
@codestergit thanks for all your work on this!
